src/client/datascience/jupyter/kernels/cellExecutionJob.ts

```diff
@@ -0,0 +1,189 @@
+    }
+    private async start() {
+        try {
+            await this.executeQueuedCells();
+            this.completion.resolve();
+        } catch (ex) {
+            // Initialize this property first, so that external users of this class know whether it has completed.
+            // Else its possible there was an error & then we wait (see next line) & in the mean time
+            // user attempts to run another cell, then `this.completion` has not completed and we end up queuing a cell
+            // but its too late.
+            // Also we in `waitForCompletion` we wait on the `this.completion` promise.
+            this.cancelledOrCompletedWithErrors = true;
+            // If something goes wrong in execution of cells or one cell, then cancel the remaining cells.
+            await this.cancel();
+            this.completion.reject(ex);
+        }
+    }
+    private async executeQueuedCells() {
+        const notebook = await this.notebookPromise;
+        this.queueOfCellsToExecute.forEach((exec) => traceCellMessage(exec.cell, 'Ready to execute'));
+        while (this.queueOfCellsToExecute.length) {
+            // Take the first item from the queue, this way we maintain order of cell executions.
+            const cellToExecute = this.queueOfCellsToExecute[0];
+            if (!cellToExecute) {
+                continue;
+            }
+            traceCellMessage(cellToExecute.cell, 'Before Execute individual cell');
+            const executionResult = await this.executeIndividualCell(cellToExecute, notebook);
+            traceCellMessage(cellToExecute.cell, `After Execute individual cell ${executionResult}`);
+            // If a cell has failed the get out.
+            if (
+                this.cancelledOrCompletedWithErrors ||
+                executionResult === vscodeNotebookEnums.NotebookCellRunState.Error
+            ) {
+                this.cancelledOrCompletedWithErrors = true;
+                traceInfo(`Cancel all remaining cells ${this.cancelledOrCompletedWithErrors} || ${executionResult}`);
+                await this.cancel();
+                break;
+            }
+            this.onCellExecutionCompleted(cellToExecute);
+        }
+    }
+    private async executeIndividualCell(
+        cellExecution: CellExecution,
+        notebook: INotebook
+    ): Promise<NotebookCellRunState | undefined> {
+        traceCellMessage(cellExecution.cell, 'Push cell into queue for execution');
+        const chainedExecution = this.chainedCellExecutionPromise.then(async () => {
+            traceCellMessage(cellExecution.cell, 'Get cell from queue for execution');
+            // Start execution
+            await cellExecution.start(notebook);
+
+            // The result promise will resolve when complete.
+            const promise = cellExecution.result;
+            promise
+                .finally(() => traceCellMessage(cellExecution.cell, 'Cell from queue completed execution'))
+                .catch(noop);
```

Comment:
Why would exceptions be swallowed?

Reply:
We return the original promise and don't swallow those.
I'm adding a finally for logging (so the original promise is still left untouched).
Will simplify this, i was just guarding against traceCellMesage failing & unnecessarily causing other errors, hut `traceCellMessage` cannot fail.

Reply:
Oh yeah duh. You're not setting the promise to the catch result.